File: src/lib/openwig/ZZone.ts
import { ZObject } from './ZObject';
import { ZCartridge } from './ZCartridge';

export class ZZone extends ZObject {
  public cartridge: ZCartridge;
  public points: { lat: number; lng: number }[] = [];
  public proximityRange: number = 10;
  public distanceRange: number = 50;
  public state: 'inside' | 'proximity' | 'distant' = 'distant';

  constructor(cartridge: ZCartridge) {
    super();
    this.cartridge = cartridge;
    this.table['Cartridge'] = cartridge.table;
    this.table['Points'] = this.points;
    this.table['ProximityRange'] = this.proximityRange;
    this.table['DistanceRange'] = this.distanceRange;
    this.table['State'] = this.state;
    
    cartridge.addObject(this);
    cartridge.zones.push(this);
  }

  public updateFromTable() {
    super.updateFromTable();
    this.proximityRange = this.table['ProximityRange'] || this.proximityRange;
    this.distanceRange = this.table['DistanceRange'] || this.distanceRange;
    this.state = this.table['State'] || this.state;
  }
}

File: java-backport/net/jakedot/jourwigo/openwig/ZZone.java
package net.jakedot.jourwigo.openwig;

import java.util.ArrayList;
import java.util.List;

public class ZZone extends ZObject {
    public ZCartridge cartridge;
    public List<Object> points = new ArrayList<>();
    public double proximityRange = 10;
    public double distanceRange = 50;
    public String state = "distant";

    public ZZone(ZCartridge cartridge) {
        super();
        this.cartridge = cartridge;
        table.put("Cartridge", cartridge.table);
        table.put("Points", points);
        table.put("ProximityRange", proximityRange);
        table.put("DistanceRange", distanceRange);
        table.put("State", state);
        
        cartridge.addObject(this);
        cartridge.zones.add(this);
    }

    @Override
    public void updateFromTable() {
        super.updateFromTable();
        if (table.containsKey("ProximityRange")) this.proximityRange = ((Number) table.get("ProximityRange")).doubleValue();
        if (table.containsKey("DistanceRange")) this.distanceRange = ((Number) table.get("DistanceRange")).doubleValue();
        if (table.containsKey("State")) this.state = (String) table.get("State");
    }
}
